snowfall/training/hmm_topo.py

```diff
@@ -0,0 +1,48 @@
+import k2
+from typing import List
+
+
+def build_hmm_topo_2state(tokens: List[int]) -> k2.Fsa:
+    """
+    Build a 2-state HMM topology used in Kaldi's chain models.
+    The first HMM state is entered only once for each token instance,
+    and the second HMM state is self-looped and optional.
+
+    Args:
+        tokens:
+            A list of token int IDs, e.g., phones, characters, etc.
+            The IDs for the first HMM state will be the same as token IDs;
+            The IDs for the second HMM state are: ``token_id + len(tokens)``
+    Returns:
+        An FST that converts a sequence of HMM state IDs to a sequence of token IDs.
+    """
+    followup_tokens = range(len(tokens), len(tokens) * 2)
+    num_states = len(tokens) + 2  # + start state, + final state
+    arcs = []
+
+    # Start state -> token state
+    for i in range(0, len(tokens)):
+        arcs += [f'0 {i + 1} {tokens[i]} {tokens[i]} 0.0']
+
+    # Token state self loops
+    for i in range(0, len(tokens)):
+        arcs += [f'{i + 1} {i + 1} {followup_tokens[i]} 0 0.0']
+
+    # Cross-token transitions
+    for i in range(0, len(tokens)):
+        for j in range(0, len(tokens)):
+            if i != j:
+                arcs += [f'{i + 1} {j + 1} {tokens[i]} {tokens[i]} 0.0']
+
+    # Token state -> superfinal state
+    for i in range(0, len(tokens)):
+        arcs += [f'{i + 1} {num_states - 1} -1 -1 0.0']
+
+    # Final state
```

Comment:
To fix the problem, you can change

```python
# Final state
arcs += [f'{num_states - 1}']

# Build the FST
arcs = '\n'.join(sorted(arcs))
```

to

```python3
# Build the FST
arcs = '\n'.join(sorted(arcs))

# Final state
arcs += f'\n{num_states - 1}'
```

---
k2 expects that the last line contains the final state.  Nothing should follow
the final state.

The documentation https://github.com/k2-fsa/k2/blob/1eeeecfac558a6ae4133e2c0b4f0022bee24c786/k2/python/k2/fsa.py#L1078
says
```python
        Caution:
          The first column has to be non-decreasing.
```

`non-decreasing` is in numeric, not in alphabetic order. `sorted` in python sorts in alphabetic.
That is the problem.

Reply:
The above fix is not a complete solution.
If the list is too large, it may result in
```
1 ....
1 ...
11 ....
2 ....
```
due to `sorted`.  11 should come after 2 and it will cause another crash.

Reply:
Changing 
```python3
arcs = '\n'.join(sorted(arcs))
```
to
```python
arcs = '\n'.join(sorted(arcs, key=lambda arc: int(arc.split()[0])))
```
should work.

Reply:
Thanks! I don't think I would have came up with that so fast myself ;)